### ISSUE TITLE
Custom encoder for data class within nested sealed class does not work

### ISSUE THREAD

**Issue by Globegitter** [sksamuel/hoplite#97]
I am having as part of my config following model:

```
data class DatabaseSecret(val host: String, val name: String, val user: String, val password: String)

sealed class Database(
    open val host: String,
    open val name: String,
    open val user: String,
    open val password: String
) {
    data class Text(
        override val host: String,
        override val name: String,
        override val user: String,
        override val password: String
    ): Database(host, name, user, password)

    data class Json(
        val content: DatabaseSecret
    ): Database(content.host, content.name, content.user, content.password)
}
```

Then I am adding a custom decoder as `content: DatabaseSecret` is provided as a json string:

```
class DatabaseSecretDecoder : NonNullableLeafDecoder<DatabaseSecret> {
    override fun supports(type: KType): Boolean = type.classifier == DatabaseSecret::class

    override fun safeLeafDecode(node: Node,
        type: KType,
        context: DecoderContext): ConfigResult<DatabaseSecret> = when (node) {
        is StringNode -> Try {
            val json = JSONObject(node.value)
            DatabaseSecret(
                host = "${json.getString("host")}:${json.getInt("port")}",
                name = json.getString("dbname"),
                user = json.getString("username"),
                password = json.getString("password")
                )
        }.toValidated { ThrowableFailure(it) }
        else -> ConfigFailure.DecodeError(node, type).invalid()
    }
}
```

This fails with the following exception:

```
Exception in thread "main" com.sksamuel.hoplite.ConfigException: Error loading config because:

    - Could not instantiate 'some.package.api.Config' because:
    
        - 'database': Could not find appropriate subclass of class some.package.api.Database: Tried some.package.api.Database$Text, some.package.api.Database$Json (/config.properties)
        at com.sksamuel.hoplite.ConfigLoader$returnOrThrow$1.invoke(ConfigLoader.kt:141)
        at com.sksamuel.hoplite.ConfigLoader$returnOrThrow$1.invoke(ConfigLoader.kt:23)
        at arrow.core.ValidatedKt.valueOr(Validated.kt:732)
        at com.sksamuel.hoplite.ConfigLoader.returnOrThrow(ConfigLoader.kt:139)
        at some.package.api.AppKt.main(App.kt:101)
        at some.package.api.AppKt.main(App.kt)
```

If however I change the `data class DatabaseSecret` to just be a class it all works as expected. Is there any reason this should not work with a data class?

**Comment by sksamuel** [comment]
It might be because the built in data class decoder is taking precedence.
I'd have to think of a way to allow users to override it. Perhaps a
priority flag that's set to low in all built in ones.

**Comment by Globegitter** [comment]
Would it not be possible to first check if there is a decode for that specific type and only if not fall back to the data class decoder?

**Comment by sksamuel** [comment]
I guess that would work too.

**Comment by sksamuel** [comment]
I've added the ability to custom decoders for data classes to automatically take precedence over the built in one. For a sealed class hierarchy, you might have to implement a custom decoder at the sealed class level.

**Comment by sksamuel** [comment]
I've released 1.2.0 where your custom decoders should take precedence over built in ones.
